### PR TITLE
Move nio ip filter rule to be a channel handler

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently nio implements ip filtering at the channel context level. This
is kind of a hack as the application logic should be implemented at the
handler level. This commit moves the ip filtering into a channel
handler. This requires adding an indicator to the channel handler to
show when a channel should be closed.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This is the first of a couple PRs that will come related to channel setup / registration. Once this is merged, I will open a PR that connects / binds on the transport thread. And one that sets socket options on the transport thread.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioHttpServerTransport.java
Comment (original-brownbear) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioHttpServerTransport.java]: Can't we just not add this handler if we don't have any IP filtering? I.e. do this:

```java
            final NioChannelHandler handler;
            if (ipFilter != null) {
                handler = new NioIPFilter(httpHandler, httpChannel.getRemoteAddress(), ipFilter, IPFilter.HTTP_PROFILE_NAME);
            } else {
                handler = httpHandler;
            }
```

the we wouldn't have to bother with the whole `null` check business in the `NioIPFilter` and save a bit of indirection?

**Review by original-brownbear — COMMENTED** [review]
Looks like a reasonable refactoring to me, just one question :)

**Comment by original-brownbear** [comment]
@tbrooks8 thanks for the changes. The test failure we're seeing is my fault but fixed in `master` => merging `master` in here should make things green :)

**Comment by tbrooks8** [comment]
@original-brownbear Thanks. I merged.

**Review by original-brownbear — APPROVED** [review]
thanks @tbrooks8 ! LGTM (assuming green CI :))

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**
